# AAA Project -  Model Building (LR, SVM, MLP)
### Katharine Cross

In [85]:
# Initial Imports 

import pandas as pd
import seaborn as sns
import numpy as np
from numpy import median
import matplotlib.pyplot as plt
import joblib
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import GridSearchCV
%matplotlib inline
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LinearRegression
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import GridSearchCV
from sklearn.ensemble import GradientBoostingClassifier
import warnings
warnings.filterwarnings('ignore', category=FutureWarning)
warnings.filterwarnings('ignore', category=DeprecationWarning)
from sklearn.svm import SVC
from sklearn.neural_network import MLPClassifier
from sklearn.metrics import accuracy_score, precision_score, recall_score
from time import time
%matplotlib inline

In [5]:
LogisticRegression()

LogisticRegression()

### Step #1: Logistic Regression with Reduced Features

In [7]:
train_features_red = pd.read_csv('train_features_reduced.csv')
train_labels = pd.read_csv('train_labels.csv')

train_features_red.head()

Average_Member_yrcost       age  count_clicks_or_opens  credit_ranges  \
0               2.070463 -0.140228               0.933098      -0.819378   
1              -0.451257 -1.272421               1.007066       1.237428   
2              -0.451257 -0.140228               0.156426      -0.819378   
3              -0.451257 -0.140228              -0.379847      -0.819378   
4              -0.451257  0.271478              -0.268894       1.237428   

   member_tenure_years  numofoffers  children_No  children_Unknown  \
0             0.255809    -0.954697    -0.493731          1.173376   
1            -0.267665    -0.954697    -0.493731         -0.852242   
2            -0.965629     0.156857    -0.493731          1.173376   
3            -0.267665    -0.954697    -0.493731          1.173376   
4             1.564492    -0.954697    -0.493731         -0.852242   

   children_Yes  dwelling_type_Multi Family Dwelling/Apartment  ...  \
0     -0.788333                                      -0.335266  ...   
1      1.268499                                      -0.335266  ...   
2     -0.788333                                      -0.335266  ...   
3     -0.788333                                      -0.335266  ...   
4      1.268499                                       2.982709  ...   

   occupation_group_Blue Collar  occupation_group_Management  \
0                     -0.248957                    -0.284569   
1                     -0.248957                    -0.284569   
2                     -0.248957                    -0.284569   
3                     -0.248957                    -0.284569   
4                     -0.248957                    -0.284569   

   occupation_group_Office Administration  occupation_group_Other  \
0                               -0.271847                0.974200   
1                               -0.271847               -1.026483   
2                               -0.271847                0.974200   
3                               -0.271847                0.974200   
4                               -0.271847                0.974200   

   occupation_group_Professional  occupation_group_Retired  \
0                      -0.361987                  -0.34641   
1                      -0.361987                  -0.34641   
2                      -0.361987                  -0.34641   
3                      -0.361987                  -0.34641   
4                      -0.361987                  -0.34641   

   occupation_group_Sales  occupation_group_Technical  member_type_Associate  \
0               -0.187528                   -0.169334              -0.442531   
1                5.332532                   -0.169334              -0.442531   
2               -0.187528                   -0.169334              -0.442531   
3               -0.187528                   -0.169334              -0.442531   
4               -0.187528                   -0.169334              -0.442531   

   member_type_Primary  
0             0.651397  
1            -1.535163  
2             0.651397  
3            -1.535163  
4             0.651397  

[5 rows x 45 columns]

In [11]:
def print_results(results):
    print('BEST PARAMS: {}\n'.format(results.best_params_))

    means = results.cv_results_['mean_test_score']
    stds = results.cv_results_['std_test_score']
    for mean, std, params in zip(means, stds, results.cv_results_['params']):
        print('{} (+/-{}) for {}'.format(round(mean, 3), round(std * 2, 3), params))

In [12]:
lr = LogisticRegression()
parameters = {
    'C': [0.001, 0.01, 0.1, 1, 10, 100, 1000]
}

cv = GridSearchCV(lr, parameters, cv=5)
cv.fit(train_features_red, train_labels.values.ravel())

print_results(cv)

BEST PARAMS: {'C': 0.1}

0.75 (+/-0.004) for {'C': 0.001}
0.758 (+/-0.025) for {'C': 0.01}
0.761 (+/-0.055) for {'C': 0.1}
0.759 (+/-0.058) for {'C': 1}
0.758 (+/-0.055) for {'C': 10}
0.758 (+/-0.055) for {'C': 100}
0.758 (+/-0.055) for {'C': 1000}


In [13]:
cv.best_estimator_

LogisticRegression(C=0.1)

In [17]:
joblib.dump(cv.best_estimator_, 'LR_model_red.pkl')

['LR_model_red.pkl']

### Step #2: Logistic Regression with Categorical Features

In [18]:
train_features_cat = pd.read_csv('train_features_cat.csv')
train_labels = pd.read_csv('train_labels.csv')

train_features_cat.head()

children_No  children_Unknown  children_Yes  \
0    -0.493731          1.173376     -0.788333   
1    -0.493731         -0.852242      1.268499   
2    -0.493731          1.173376     -0.788333   
3    -0.493731          1.173376     -0.788333   
4    -0.493731         -0.852242      1.268499   

   dwelling_type_Multi Family Dwelling/Apartment  dwelling_type_PO Box  \
0                                      -0.335266             -0.139779   
1                                      -0.335266             -0.139779   
2                                      -0.335266             -0.139779   
3                                      -0.335266             -0.139779   
4                                       2.982709             -0.139779   

   dwelling_type_Single Family  dwelling_type_Unknown  \
0                    -0.921212               1.173376   
1                     1.085527              -0.852242   
2                    -0.921212               1.173376   
3                    -0.921212               1.173376   
4                    -0.921212              -0.852242   

   education_Completed College  education_Graduate School  \
0                    -0.280976                  -0.107022   
1                    -0.280976                  -0.107022   
2                    -0.280976                  -0.107022   
3                    -0.280976                  -0.107022   
4                     3.559026                  -0.107022   

   education_High School  ...  occupation_group_Management  \
0              -0.400959  ...                    -0.284569   
1              -0.400959  ...                    -0.284569   
2              -0.400959  ...                    -0.284569   
3              -0.400959  ...                    -0.284569   
4              -0.400959  ...                    -0.284569   

   occupation_group_Office Administration  occupation_group_Other  \
0                               -0.271847                0.974200   
1                               -0.271847               -1.026483   
2                               -0.271847                0.974200   
3                               -0.271847                0.974200   
4                               -0.271847                0.974200   

   occupation_group_Professional  occupation_group_Retired  \
0                      -0.361987                  -0.34641   
1                      -0.361987                  -0.34641   
2                      -0.361987                  -0.34641   
3                      -0.361987                  -0.34641   
4                      -0.361987                  -0.34641   

   occupation_group_Sales  occupation_group_Technical  member_type_Associate  \
0               -0.187528                   -0.169334              -0.442531   
1                5.332532                   -0.169334              -0.442531   
2               -0.187528                   -0.169334              -0.442531   
3               -0.187528                   -0.169334              -0.442531   
4               -0.187528                   -0.169334              -0.442531   

   member_type_Primary  member_type_Unknown  
0             0.651397            -0.393611  
1            -1.535163             2.540580  
2             0.651397            -0.393611  
3            -1.535163             2.540580  
4             0.651397            -0.393611  

[5 rows x 80 columns]

In [19]:
lr = LogisticRegression()
parameters = {
    'C': [0.001, 0.01, 0.1, 1, 10, 100, 1000]
}

cv = GridSearchCV(lr, parameters, cv=5)
cv.fit(train_features_cat, train_labels.values.ravel())

print_results(cv)

C:\Users\student\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


BEST PARAMS: {'C': 0.01}

0.75 (+/-0.004) for {'C': 0.001}
0.752 (+/-0.017) for {'C': 0.01}
0.745 (+/-0.048) for {'C': 0.1}
0.742 (+/-0.053) for {'C': 1}
0.742 (+/-0.05) for {'C': 10}
0.742 (+/-0.05) for {'C': 100}
0.742 (+/-0.05) for {'C': 1000}


In [20]:
cv.best_estimator_

LogisticRegression(C=0.01)

In [21]:
joblib.dump(cv.best_estimator_, 'LR_model_cat.pkl')

['LR_model_cat.pkl']

### Step #3: Logistic Regression with Numerical Features

In [23]:
train_features_num = pd.read_csv('train_features_num.csv')
train_labels = pd.read_csv('train_labels.csv')

train_features_num.head()

Average_Member_yrcost       age  count_clicks_or_opens  credit_ranges  \
0               2.070463 -0.140228               0.933098      -0.819378   
1              -0.451257 -1.272421               1.007066       1.237428   
2              -0.451257 -0.140228               0.156426      -0.819378   
3              -0.451257 -0.140228              -0.379847      -0.819378   
4              -0.451257  0.271478              -0.268894       1.237428   

   member_tenure_years  numofoffers  
0             0.255809    -0.954697  
1            -0.267665    -0.954697  
2            -0.965629     0.156857  
3            -0.267665    -0.954697  
4             1.564492    -0.954697

In [28]:
lr = LogisticRegression()
parameters = {
    'C': [0.001, 0.01, 0.1, 1, 10, 100, 1000]
}

cv = GridSearchCV(lr, parameters, cv=5)
cv.fit(train_features_num, train_labels.values.ravel())

print_results(cv)

BEST PARAMS: {'C': 1}

0.75 (+/-0.004) for {'C': 0.001}
0.752 (+/-0.01) for {'C': 0.01}
0.754 (+/-0.028) for {'C': 0.1}
0.757 (+/-0.026) for {'C': 1}
0.756 (+/-0.029) for {'C': 10}
0.756 (+/-0.029) for {'C': 100}
0.756 (+/-0.029) for {'C': 1000}


In [25]:
cv.best_estimator_

LogisticRegression(C=1)

In [26]:
joblib.dump(cv.best_estimator_, 'LR_model_num.pkl')

['LR_model_num.pkl']

### Step #4: Logistic Regression with All Features

In [27]:
train_features_all = pd.read_csv('train_features_all.csv')
train_labels = pd.read_csv('train_labels.csv')

train_features_all.head()

Average_Member_yrcost       age  count_clicks_or_opens  credit_ranges  \
0               2.070463 -0.140228               0.933098      -0.819378   
1              -0.451257 -1.272421               1.007066       1.237428   
2              -0.451257 -0.140228               0.156426      -0.819378   
3              -0.451257 -0.140228              -0.379847      -0.819378   
4              -0.451257  0.271478              -0.268894       1.237428   

   member_tenure_years  numofoffers  children_No  children_Unknown  \
0             0.255809    -0.954697    -0.493731          1.173376   
1            -0.267665    -0.954697    -0.493731         -0.852242   
2            -0.965629     0.156857    -0.493731          1.173376   
3            -0.267665    -0.954697    -0.493731          1.173376   
4             1.564492    -0.954697    -0.493731         -0.852242   

   children_Yes  dwelling_type_Multi Family Dwelling/Apartment  ...  \
0     -0.788333                                      -0.335266  ...   
1      1.268499                                      -0.335266  ...   
2     -0.788333                                      -0.335266  ...   
3     -0.788333                                      -0.335266  ...   
4      1.268499                                       2.982709  ...   

   occupation_group_Management  occupation_group_Office Administration  \
0                    -0.284569                               -0.271847   
1                    -0.284569                               -0.271847   
2                    -0.284569                               -0.271847   
3                    -0.284569                               -0.271847   
4                    -0.284569                               -0.271847   

   occupation_group_Other  occupation_group_Professional  \
0                0.974200                      -0.361987   
1               -1.026483                      -0.361987   
2                0.974200                      -0.361987   
3                0.974200                      -0.361987   
4                0.974200                      -0.361987   

   occupation_group_Retired  occupation_group_Sales  \
0                  -0.34641               -0.187528   
1                  -0.34641                5.332532   
2                  -0.34641               -0.187528   
3                  -0.34641               -0.187528   
4                  -0.34641               -0.187528   

   occupation_group_Technical  member_type_Associate  member_type_Primary  \
0                   -0.169334              -0.442531             0.651397   
1                   -0.169334              -0.442531            -1.535163   
2                   -0.169334              -0.442531             0.651397   
3                   -0.169334              -0.442531            -1.535163   
4                   -0.169334              -0.442531             0.651397   

   member_type_Unknown  
0            -0.393611  
1             2.540580  
2            -0.393611  
3             2.540580  
4            -0.393611  

[5 rows x 86 columns]

In [29]:
lr = LogisticRegression()
parameters = {
    'C': [0.001, 0.01, 0.1, 1, 10, 100, 1000]
}

cv = GridSearchCV(lr, parameters, cv=5)
cv.fit(train_features_all, train_labels.values.ravel())

print_results(cv)

C:\Users\student\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


BEST PARAMS: {'C': 0.01}

0.75 (+/-0.004) for {'C': 0.001}
0.762 (+/-0.01) for {'C': 0.01}
0.759 (+/-0.036) for {'C': 0.1}
0.754 (+/-0.058) for {'C': 1}
0.755 (+/-0.06) for {'C': 10}
0.755 (+/-0.06) for {'C': 100}
0.755 (+/-0.06) for {'C': 1000}


In [30]:
cv.best_estimator_

LogisticRegression(C=0.01)

In [31]:
joblib.dump(cv.best_estimator_, 'LR_model_all.pkl')

['LR_model_all.pkl']

### Step #5: Support Vector Machine with Reduced Features

In [34]:
SVC()

SVC()

In [35]:
svc = SVC()
parameters = {
    'kernel': ['linear', 'rbf'],
    'C': [0.1, 1, 10]
}

cv = GridSearchCV(svc, parameters, cv=5)
cv.fit(train_features_red, train_labels.values.ravel())

print_results(cv)

BEST PARAMS: {'C': 0.1, 'kernel': 'rbf'}

0.747 (+/-0.007) for {'C': 0.1, 'kernel': 'linear'}
0.75 (+/-0.004) for {'C': 0.1, 'kernel': 'rbf'}
0.747 (+/-0.007) for {'C': 1, 'kernel': 'linear'}
0.749 (+/-0.008) for {'C': 1, 'kernel': 'rbf'}
0.747 (+/-0.007) for {'C': 10, 'kernel': 'linear'}
0.699 (+/-0.076) for {'C': 10, 'kernel': 'rbf'}


In [36]:
cv.best_estimator_

SVC(C=0.1)

In [37]:
joblib.dump(cv.best_estimator_, 'SVM_model_red.pkl')

['SVM_model_red.pkl']

### Step #6: Support Vector Machine with Categorical Features

In [38]:
svc = SVC()
parameters = {
    'kernel': ['linear', 'rbf'],
    'C': [0.1, 1, 10]
}

cv = GridSearchCV(svc, parameters, cv=5)
cv.fit(train_features_cat, train_labels.values.ravel())

print_results(cv)

BEST PARAMS: {'C': 0.1, 'kernel': 'rbf'}

0.746 (+/-0.017) for {'C': 0.1, 'kernel': 'linear'}
0.75 (+/-0.004) for {'C': 0.1, 'kernel': 'rbf'}
0.746 (+/-0.017) for {'C': 1, 'kernel': 'linear'}
0.748 (+/-0.008) for {'C': 1, 'kernel': 'rbf'}
0.746 (+/-0.017) for {'C': 10, 'kernel': 'linear'}
0.697 (+/-0.048) for {'C': 10, 'kernel': 'rbf'}


In [39]:
cv.best_estimator_

SVC(C=0.1)

In [40]:
joblib.dump(cv.best_estimator_, 'SVM_model_cat.pkl')

['SVM_model_cat.pkl']

### Step #7: Support Vector Machine with Numerical Features

In [44]:
svc = SVC()
parameters = {
    'kernel': ['linear', 'rbf'],
    'C': [0.1, 1, 10]
}

cv = GridSearchCV(svc, parameters, cv=5)
cv.fit(train_features_num, train_labels.values.ravel())

print_results(cv)

BEST PARAMS: {'C': 10, 'kernel': 'rbf'}

0.75 (+/-0.004) for {'C': 0.1, 'kernel': 'linear'}
0.75 (+/-0.004) for {'C': 0.1, 'kernel': 'rbf'}
0.75 (+/-0.004) for {'C': 1, 'kernel': 'linear'}
0.75 (+/-0.009) for {'C': 1, 'kernel': 'rbf'}
0.75 (+/-0.004) for {'C': 10, 'kernel': 'linear'}
0.751 (+/-0.031) for {'C': 10, 'kernel': 'rbf'}


In [42]:
cv.best_estimator_

SVC(C=10)

In [43]:
joblib.dump(cv.best_estimator_, 'SVM_model_num.pkl')

['SVM_model_num.pkl']

### Step #8: Support Vector Machine with All Features

In [45]:
svc = SVC()
parameters = {
    'kernel': ['linear', 'rbf'],
    'C': [0.1, 1, 10]
}

cv = GridSearchCV(svc, parameters, cv=5)
cv.fit(train_features_all, train_labels.values.ravel())

print_results(cv)

BEST PARAMS: {'C': 10, 'kernel': 'linear'}

0.758 (+/-0.027) for {'C': 0.1, 'kernel': 'linear'}
0.75 (+/-0.004) for {'C': 0.1, 'kernel': 'rbf'}
0.76 (+/-0.029) for {'C': 1, 'kernel': 'linear'}
0.75 (+/-0.006) for {'C': 1, 'kernel': 'rbf'}
0.763 (+/-0.026) for {'C': 10, 'kernel': 'linear'}
0.707 (+/-0.048) for {'C': 10, 'kernel': 'rbf'}


In [46]:
cv.best_estimator_

SVC(C=10, kernel='linear')

In [47]:
joblib.dump(cv.best_estimator_, 'SVM_model_all.pkl')

['SVM_model_all.pkl']

### Step #9: Multi-Layer Perceptron with Reduced Features

In [49]:
mlp = MLPClassifier()
parameters = {
    'hidden_layer_sizes': [(10,), (50,), (100,)],
    'activation': ['relu', 'tanh', 'logistic'],
    'learning_rate': ['constant', 'invscaling', 'adaptive']
}

cv = GridSearchCV(mlp, parameters, cv=5)
cv.fit(train_features_red, train_labels.values.ravel())

print_results(cv)

C:\Users\student\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
C:\Users\student\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
C:\Users\student\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
C:\Users\student\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
C:\Users\student\anaconda3\lib\site-packages\sklearn\neural_netw

C:\Users\student\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
C:\Users\student\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
C:\Users\student\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
C:\Users\student\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
C:\Users\student\anaconda3\lib\site-packages\sklearn\neural_netw

C:\Users\student\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
C:\Users\student\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
C:\Users\student\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
C:\Users\student\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
C:\Users\student\anaconda3\lib\site-packages\sklearn\neural_netw

C:\Users\student\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
C:\Users\student\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
C:\Users\student\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
C:\Users\student\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
C:\Users\student\anaconda3\lib\site-packages\sklearn\neural_netw

BEST PARAMS: {'activation': 'logistic', 'hidden_layer_sizes': (10,), 'learning_rate': 'adaptive'}

0.733 (+/-0.056) for {'activation': 'relu', 'hidden_layer_sizes': (10,), 'learning_rate': 'constant'}
0.742 (+/-0.015) for {'activation': 'relu', 'hidden_layer_sizes': (10,), 'learning_rate': 'invscaling'}
0.739 (+/-0.028) for {'activation': 'relu', 'hidden_layer_sizes': (10,), 'learning_rate': 'adaptive'}
0.713 (+/-0.025) for {'activation': 'relu', 'hidden_layer_sizes': (50,), 'learning_rate': 'constant'}
0.726 (+/-0.021) for {'activation': 'relu', 'hidden_layer_sizes': (50,), 'learning_rate': 'invscaling'}
0.708 (+/-0.014) for {'activation': 'relu', 'hidden_layer_sizes': (50,), 'learning_rate': 'adaptive'}
0.705 (+/-0.025) for {'activation': 'relu', 'hidden_layer_sizes': (100,), 'learning_rate': 'constant'}
0.719 (+/-0.042) for {'activation': 'relu', 'hidden_layer_sizes': (100,), 'learning_rate': 'invscaling'}
0.707 (+/-0.038) for {'activation': 'relu', 'hidden_layer_sizes': (100,), 'le

C:\Users\student\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


In [50]:
cv.best_estimator_

MLPClassifier(activation='logistic', hidden_layer_sizes=(10,),
              learning_rate='adaptive')

In [51]:
joblib.dump(cv.best_estimator_, 'MLP_model_red.pkl')

['MLP_model_red.pkl']

### Step #10: Multi-Layer Perceptron with Categorical Features

In [52]:
mlp = MLPClassifier()
parameters = {
    'hidden_layer_sizes': [(10,), (50,), (100,)],
    'activation': ['relu', 'tanh', 'logistic'],
    'learning_rate': ['constant', 'invscaling', 'adaptive']
}

cv = GridSearchCV(mlp, parameters, cv=5)
cv.fit(train_features_cat, train_labels.values.ravel())

print_results(cv)

C:\Users\student\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
C:\Users\student\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
C:\Users\student\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
C:\Users\student\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
C:\Users\student\anaconda3\lib\site-packages\sklearn\neural_netw

C:\Users\student\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
C:\Users\student\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
C:\Users\student\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
C:\Users\student\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
C:\Users\student\anaconda3\lib\site-packages\sklearn\neural_netw

C:\Users\student\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
C:\Users\student\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
C:\Users\student\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
C:\Users\student\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
C:\Users\student\anaconda3\lib\site-packages\sklearn\neural_netw

C:\Users\student\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
C:\Users\student\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
C:\Users\student\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
C:\Users\student\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
C:\Users\student\anaconda3\lib\site-packages\sklearn\neural_netw

BEST PARAMS: {'activation': 'logistic', 'hidden_layer_sizes': (10,), 'learning_rate': 'constant'}

0.707 (+/-0.055) for {'activation': 'relu', 'hidden_layer_sizes': (10,), 'learning_rate': 'constant'}
0.693 (+/-0.036) for {'activation': 'relu', 'hidden_layer_sizes': (10,), 'learning_rate': 'invscaling'}
0.696 (+/-0.035) for {'activation': 'relu', 'hidden_layer_sizes': (10,), 'learning_rate': 'adaptive'}
0.666 (+/-0.057) for {'activation': 'relu', 'hidden_layer_sizes': (50,), 'learning_rate': 'constant'}
0.685 (+/-0.032) for {'activation': 'relu', 'hidden_layer_sizes': (50,), 'learning_rate': 'invscaling'}
0.688 (+/-0.042) for {'activation': 'relu', 'hidden_layer_sizes': (50,), 'learning_rate': 'adaptive'}
0.686 (+/-0.055) for {'activation': 'relu', 'hidden_layer_sizes': (100,), 'learning_rate': 'constant'}
0.686 (+/-0.065) for {'activation': 'relu', 'hidden_layer_sizes': (100,), 'learning_rate': 'invscaling'}
0.675 (+/-0.044) for {'activation': 'relu', 'hidden_layer_sizes': (100,), 'le

C:\Users\student\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


In [54]:
cv.best_estimator_

MLPClassifier(activation='logistic', hidden_layer_sizes=(10,))

In [55]:
joblib.dump(cv.best_estimator_, 'MLP_model_cat.pkl')

['MLP_model_cat.pkl']

### Step #11: Multi-Layer Perceptron with Numerical Features


In [56]:
mlp = MLPClassifier()
parameters = {
    'hidden_layer_sizes': [(10,), (50,), (100,)],
    'activation': ['relu', 'tanh', 'logistic'],
    'learning_rate': ['constant', 'invscaling', 'adaptive']
}

cv = GridSearchCV(mlp, parameters, cv=5)
cv.fit(train_features_num, train_labels.values.ravel())

print_results(cv)

C:\Users\student\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
C:\Users\student\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
C:\Users\student\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
C:\Users\student\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
C:\Users\student\anaconda3\lib\site-packages\sklearn\neural_netw

C:\Users\student\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
C:\Users\student\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
C:\Users\student\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
C:\Users\student\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
C:\Users\student\anaconda3\lib\site-packages\sklearn\neural_netw

BEST PARAMS: {'activation': 'tanh', 'hidden_layer_sizes': (10,), 'learning_rate': 'adaptive'}

0.753 (+/-0.025) for {'activation': 'relu', 'hidden_layer_sizes': (10,), 'learning_rate': 'constant'}
0.752 (+/-0.022) for {'activation': 'relu', 'hidden_layer_sizes': (10,), 'learning_rate': 'invscaling'}
0.755 (+/-0.018) for {'activation': 'relu', 'hidden_layer_sizes': (10,), 'learning_rate': 'adaptive'}
0.747 (+/-0.023) for {'activation': 'relu', 'hidden_layer_sizes': (50,), 'learning_rate': 'constant'}
0.756 (+/-0.023) for {'activation': 'relu', 'hidden_layer_sizes': (50,), 'learning_rate': 'invscaling'}
0.753 (+/-0.026) for {'activation': 'relu', 'hidden_layer_sizes': (50,), 'learning_rate': 'adaptive'}
0.751 (+/-0.023) for {'activation': 'relu', 'hidden_layer_sizes': (100,), 'learning_rate': 'constant'}
0.751 (+/-0.024) for {'activation': 'relu', 'hidden_layer_sizes': (100,), 'learning_rate': 'invscaling'}
0.753 (+/-0.021) for {'activation': 'relu', 'hidden_layer_sizes': (100,), 'learni

In [58]:
cv.best_estimator_

MLPClassifier(activation='tanh', hidden_layer_sizes=(10,),
              learning_rate='adaptive')

In [59]:
joblib.dump(cv.best_estimator_, 'MLP_model_num.pkl')

['MLP_model_num.pkl']

### Step #12: Multi-Layer Perceptron with All Features


In [60]:
mlp = MLPClassifier()
parameters = {
    'hidden_layer_sizes': [(10,), (50,), (100,)],
    'activation': ['relu', 'tanh', 'logistic'],
    'learning_rate': ['constant', 'invscaling', 'adaptive']
}

cv = GridSearchCV(mlp, parameters, cv=5)
cv.fit(train_features_all, train_labels.values.ravel())

print_results(cv)

C:\Users\student\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
C:\Users\student\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
C:\Users\student\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
C:\Users\student\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
C:\Users\student\anaconda3\lib\site-packages\sklearn\neural_netw

C:\Users\student\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
C:\Users\student\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
C:\Users\student\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
C:\Users\student\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
C:\Users\student\anaconda3\lib\site-packages\sklearn\neural_netw

C:\Users\student\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
C:\Users\student\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
C:\Users\student\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
C:\Users\student\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
C:\Users\student\anaconda3\lib\site-packages\sklearn\neural_netw

C:\Users\student\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
C:\Users\student\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
C:\Users\student\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
C:\Users\student\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
C:\Users\student\anaconda3\lib\site-packages\sklearn\neural_netw

BEST PARAMS: {'activation': 'logistic', 'hidden_layer_sizes': (10,), 'learning_rate': 'adaptive'}

0.724 (+/-0.047) for {'activation': 'relu', 'hidden_layer_sizes': (10,), 'learning_rate': 'constant'}
0.72 (+/-0.062) for {'activation': 'relu', 'hidden_layer_sizes': (10,), 'learning_rate': 'invscaling'}
0.721 (+/-0.03) for {'activation': 'relu', 'hidden_layer_sizes': (10,), 'learning_rate': 'adaptive'}
0.711 (+/-0.064) for {'activation': 'relu', 'hidden_layer_sizes': (50,), 'learning_rate': 'constant'}
0.707 (+/-0.079) for {'activation': 'relu', 'hidden_layer_sizes': (50,), 'learning_rate': 'invscaling'}
0.712 (+/-0.054) for {'activation': 'relu', 'hidden_layer_sizes': (50,), 'learning_rate': 'adaptive'}
0.713 (+/-0.06) for {'activation': 'relu', 'hidden_layer_sizes': (100,), 'learning_rate': 'constant'}
0.716 (+/-0.046) for {'activation': 'relu', 'hidden_layer_sizes': (100,), 'learning_rate': 'invscaling'}
0.716 (+/-0.049) for {'activation': 'relu', 'hidden_layer_sizes': (100,), 'learn

C:\Users\student\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


In [63]:
cv.best_estimator_

MLPClassifier(activation='logistic', hidden_layer_sizes=(10,),
              learning_rate='adaptive')

In [64]:
joblib.dump(cv.best_estimator_, 'MLP_model_all.pkl')

['MLP_model_all.pkl']

### Step #13: Gradient Boosting with Reduced Features


In [65]:
gb = GradientBoostingClassifier()
parameters = {
    'n_estimators': [5, 50, 250, 500],
    'max_depth': [1, 3, 5, 7, 9],
    'learning_rate': [0.01, 0.1, 1, 10, 100]
}

cv = GridSearchCV(gb, parameters, cv=5)
cv.fit(train_features_red, train_labels.values.ravel())

print_results(cv)

BEST PARAMS: {'learning_rate': 0.01, 'max_depth': 3, 'n_estimators': 250}

0.75 (+/-0.004) for {'learning_rate': 0.01, 'max_depth': 1, 'n_estimators': 5}
0.75 (+/-0.004) for {'learning_rate': 0.01, 'max_depth': 1, 'n_estimators': 50}
0.751 (+/-0.006) for {'learning_rate': 0.01, 'max_depth': 1, 'n_estimators': 250}
0.754 (+/-0.008) for {'learning_rate': 0.01, 'max_depth': 1, 'n_estimators': 500}
0.75 (+/-0.004) for {'learning_rate': 0.01, 'max_depth': 3, 'n_estimators': 5}
0.75 (+/-0.004) for {'learning_rate': 0.01, 'max_depth': 3, 'n_estimators': 50}
0.76 (+/-0.02) for {'learning_rate': 0.01, 'max_depth': 3, 'n_estimators': 250}
0.753 (+/-0.022) for {'learning_rate': 0.01, 'max_depth': 3, 'n_estimators': 500}
0.75 (+/-0.004) for {'learning_rate': 0.01, 'max_depth': 5, 'n_estimators': 5}
0.744 (+/-0.006) for {'learning_rate': 0.01, 'max_depth': 5, 'n_estimators': 50}
0.749 (+/-0.021) for {'learning_rate': 0.01, 'max_depth': 5, 'n_estimators': 250}
0.751 (+/-0.022) for {'learning_rate': 

In [66]:
cv.best_estimator_

GradientBoostingClassifier(learning_rate=0.01, n_estimators=250)

In [67]:
joblib.dump(cv.best_estimator_, 'GB_model_red.pkl')

['GB_model_red.pkl']

### Step #14: Gradient Boosting with Categorical Features


In [68]:
gb = GradientBoostingClassifier()
parameters = {
    'n_estimators': [5, 50, 250, 500],
    'max_depth': [1, 3, 5, 7, 9],
    'learning_rate': [0.01, 0.1, 1, 10, 100]
}

cv = GridSearchCV(gb, parameters, cv=5)
cv.fit(train_features_cat, train_labels.values.ravel())

print_results(cv)

BEST PARAMS: {'learning_rate': 0.1, 'max_depth': 1, 'n_estimators': 250}

0.75 (+/-0.004) for {'learning_rate': 0.01, 'max_depth': 1, 'n_estimators': 5}
0.75 (+/-0.004) for {'learning_rate': 0.01, 'max_depth': 1, 'n_estimators': 50}
0.75 (+/-0.004) for {'learning_rate': 0.01, 'max_depth': 1, 'n_estimators': 250}
0.749 (+/-0.006) for {'learning_rate': 0.01, 'max_depth': 1, 'n_estimators': 500}
0.75 (+/-0.004) for {'learning_rate': 0.01, 'max_depth': 3, 'n_estimators': 5}
0.75 (+/-0.004) for {'learning_rate': 0.01, 'max_depth': 3, 'n_estimators': 50}
0.747 (+/-0.011) for {'learning_rate': 0.01, 'max_depth': 3, 'n_estimators': 250}
0.744 (+/-0.014) for {'learning_rate': 0.01, 'max_depth': 3, 'n_estimators': 500}
0.75 (+/-0.004) for {'learning_rate': 0.01, 'max_depth': 5, 'n_estimators': 5}
0.748 (+/-0.006) for {'learning_rate': 0.01, 'max_depth': 5, 'n_estimators': 50}
0.733 (+/-0.02) for {'learning_rate': 0.01, 'max_depth': 5, 'n_estimators': 250}
0.726 (+/-0.028) for {'learning_rate': 0

In [69]:
cv.best_estimator_

GradientBoostingClassifier(max_depth=1, n_estimators=250)

In [70]:
joblib.dump(cv.best_estimator_, 'GB_model_cat.pkl')

['GB_model_cat.pkl']

### Step #15: Gradient Boosting with Numerical Features


In [71]:
gb = GradientBoostingClassifier()
parameters = {
    'n_estimators': [5, 50, 250, 500],
    'max_depth': [1, 3, 5, 7, 9],
    'learning_rate': [0.01, 0.1, 1, 10, 100]
}

cv = GridSearchCV(gb, parameters, cv=5)
cv.fit(train_features_num, train_labels.values.ravel())

print_results(cv)

BEST PARAMS: {'learning_rate': 1, 'max_depth': 1, 'n_estimators': 5}

0.75 (+/-0.004) for {'learning_rate': 0.01, 'max_depth': 1, 'n_estimators': 5}
0.75 (+/-0.004) for {'learning_rate': 0.01, 'max_depth': 1, 'n_estimators': 50}
0.75 (+/-0.007) for {'learning_rate': 0.01, 'max_depth': 1, 'n_estimators': 250}
0.75 (+/-0.008) for {'learning_rate': 0.01, 'max_depth': 1, 'n_estimators': 500}
0.75 (+/-0.004) for {'learning_rate': 0.01, 'max_depth': 3, 'n_estimators': 5}
0.75 (+/-0.004) for {'learning_rate': 0.01, 'max_depth': 3, 'n_estimators': 50}
0.746 (+/-0.009) for {'learning_rate': 0.01, 'max_depth': 3, 'n_estimators': 250}
0.75 (+/-0.012) for {'learning_rate': 0.01, 'max_depth': 3, 'n_estimators': 500}
0.75 (+/-0.004) for {'learning_rate': 0.01, 'max_depth': 5, 'n_estimators': 5}
0.752 (+/-0.015) for {'learning_rate': 0.01, 'max_depth': 5, 'n_estimators': 50}
0.751 (+/-0.031) for {'learning_rate': 0.01, 'max_depth': 5, 'n_estimators': 250}
0.74 (+/-0.032) for {'learning_rate': 0.01, '

In [72]:
cv.best_estimator_

GradientBoostingClassifier(learning_rate=1, max_depth=1, n_estimators=5)

In [73]:
joblib.dump(cv.best_estimator_, 'GB_model_num.pkl')

['GB_model_num.pkl']

### Step #15: Gradient Boosting with All Features


In [74]:
gb = GradientBoostingClassifier()
parameters = {
    'n_estimators': [5, 50, 250, 500],
    'max_depth': [1, 3, 5, 7, 9],
    'learning_rate': [0.01, 0.1, 1, 10, 100]
}

cv = GridSearchCV(gb, parameters, cv=5)
cv.fit(train_features_all, train_labels.values.ravel())

print_results(cv)

BEST PARAMS: {'learning_rate': 0.1, 'max_depth': 1, 'n_estimators': 500}

0.75 (+/-0.004) for {'learning_rate': 0.01, 'max_depth': 1, 'n_estimators': 5}
0.75 (+/-0.004) for {'learning_rate': 0.01, 'max_depth': 1, 'n_estimators': 50}
0.751 (+/-0.006) for {'learning_rate': 0.01, 'max_depth': 1, 'n_estimators': 250}
0.752 (+/-0.006) for {'learning_rate': 0.01, 'max_depth': 1, 'n_estimators': 500}
0.75 (+/-0.004) for {'learning_rate': 0.01, 'max_depth': 3, 'n_estimators': 5}
0.75 (+/-0.004) for {'learning_rate': 0.01, 'max_depth': 3, 'n_estimators': 50}
0.754 (+/-0.01) for {'learning_rate': 0.01, 'max_depth': 3, 'n_estimators': 250}
0.741 (+/-0.027) for {'learning_rate': 0.01, 'max_depth': 3, 'n_estimators': 500}
0.75 (+/-0.004) for {'learning_rate': 0.01, 'max_depth': 5, 'n_estimators': 5}
0.747 (+/-0.01) for {'learning_rate': 0.01, 'max_depth': 5, 'n_estimators': 50}
0.743 (+/-0.042) for {'learning_rate': 0.01, 'max_depth': 5, 'n_estimators': 250}
0.744 (+/-0.051) for {'learning_rate': 0

In [75]:
cv.best_estimator_

GradientBoostingClassifier(max_depth=1, n_estimators=500)

In [76]:
joblib.dump(cv.best_estimator_, 'GB_model_all.pkl')

['GB_model_all.pkl']

### Step #16: Evaluate Every Model on the Test Set


In [80]:
val_features_cat = pd.read_csv('val_features_cat.csv')
val_features_num = pd.read_csv('val_features_num.csv')
val_features_all = pd.read_csv('val_features_all.csv')
val_features_red = pd.read_csv('val_features_reduced.csv')

val_labels = pd.read_csv('val_labels.csv')

val_features_red.head()

Average_Member_yrcost       age  count_clicks_or_opens  credit_ranges  \
0               1.097030 -0.090995              -0.379847      -0.819378   
1              -0.451257  2.330009              -0.379847      -0.819378   
2              -0.451257 -2.713393              -0.379847      -0.819378   
3              -0.451257  2.844642              -0.379847       1.237428   
4               0.402632 -2.404613              -0.379847       1.237428   

   member_tenure_years  numofoffers  children_No  children_Unknown  \
0            -0.703892     0.156857     2.025394         -0.852242   
1            -0.180419     0.156857     2.025394         -0.852242   
2            -0.267665    -0.954697    -0.493731         -0.852242   
3             1.302756     1.268411    -0.493731         -0.852242   
4             0.517546    -0.954697    -0.493731         -0.852242   

   children_Yes  dwelling_type_Multi Family Dwelling/Apartment  ...  \
0     -0.788333                                      -0.335266  ...   
1     -0.788333                                      -0.335266  ...   
2      1.268499                                      -0.335266  ...   
3      1.268499                                      -0.335266  ...   
4      1.268499                                      -0.335266  ...   

   occupation_group_Blue Collar  occupation_group_Management  \
0                     -0.248957                    -0.284569   
1                     -0.248957                    -0.284569   
2                     -0.248957                    -0.284569   
3                     -0.248957                    -0.284569   
4                     -0.248957                     3.514091   

   occupation_group_Office Administration  occupation_group_Other  \
0                               -0.271847                0.974200   
1                               -0.271847                0.974200   
2                               -0.271847               -1.026483   
3                               -0.271847               -1.026483   
4                               -0.271847               -1.026483   

   occupation_group_Professional  occupation_group_Retired  \
0                      -0.361987                 -0.346410   
1                      -0.361987                 -0.346410   
2                      -0.361987                  2.886751   
3                      -0.361987                  2.886751   
4                      -0.361987                 -0.346410   

   occupation_group_Sales  occupation_group_Technical  member_type_Associate  \
0               -0.187528                   -0.169334              -0.442531   
1               -0.187528                   -0.169334              -0.442531   
2               -0.187528                   -0.169334              -0.442531   
3               -0.187528                   -0.169334              -0.442531   
4               -0.187528                   -0.169334               2.259731   

   member_type_Primary  
0             0.651397  
1             0.651397  
2            -1.535163  
3             0.651397  
4            -1.535163  

[5 rows x 45 columns]

#### A) Random Forest

In [81]:
# Read in models
models = {}

for mdl in ['cat', 'num', 'all', 'red']:
    models[mdl] = joblib.load('rf_features_{}.pkl'.format(mdl))

In [82]:
def evaluate_model(name, model, features, labels):
    start = time()
    pred = model.predict(features)
    end = time()
    accuracy = round(accuracy_score(labels, pred), 3)
    precision = round(precision_score(labels, pred), 3)
    recall = round(recall_score(labels, pred), 3)
    print('{} -- \tAccuracy: {} / Precision: {} / Recall: {} / Latency: {}ms'.format(name,
                                                                                     accuracy,
                                                                                     precision,
                                                                                     recall,
                                                                                     round((end - start)*1000, 1)))

In [86]:
# Evaluate all of our models on the validation set
evaluate_model('Categorical Features - Random Forest', models['cat'], val_features_cat, val_labels)
evaluate_model('Numerical Features - Random Forest', models['num'], val_features_num, val_labels)
evaluate_model('All Features - Random Forest', models['all'], val_features_all, val_labels)
evaluate_model('Reduced Features - Random Forest', models['red'], val_features_red, val_labels)

Categorical Features - Random Forest -- 	Accuracy: 0.768 / Precision: 0.0 / Recall: 0.0 / Latency: 6.2ms
Numerical Features - Random Forest -- 	Accuracy: 0.756 / Precision: 0.412 / Recall: 0.123 / Latency: 9.1ms
All Features - Random Forest -- 	Accuracy: 0.78 / Precision: 0.636 / Recall: 0.123 / Latency: 3.0ms
Reduced Features - Random Forest -- 	Accuracy: 0.772 / Precision: 0.6 / Recall: 0.053 / Latency: 10.0ms


C:\Users\student\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1318: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 due to no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


#### B) Logistic Regression

In [88]:
# Read in models
models = {}

for mdl in ['cat', 'num', 'all', 'red']:
    models[mdl] = joblib.load('LR_model_{}.pkl'.format(mdl))

In [89]:
# Evaluate all of our models on the validation set
evaluate_model('Categorical Features - Logistic Regression', models['cat'], val_features_cat, val_labels)
evaluate_model('Numerical Features - Logistic Regression', models['num'], val_features_num, val_labels)
evaluate_model('All Features - Logistic Regression', models['all'], val_features_all, val_labels)
evaluate_model('Reduced Features - Logistic Regression', models['red'], val_features_red, val_labels)

Categorical Features - Logistic Regression -- 	Accuracy: 0.768 / Precision: 0.5 / Recall: 0.018 / Latency: 3.0ms
Numerical Features - Logistic Regression -- 	Accuracy: 0.785 / Precision: 0.7 / Recall: 0.123 / Latency: 1.0ms
All Features - Logistic Regression -- 	Accuracy: 0.752 / Precision: 0.333 / Recall: 0.07 / Latency: 2.0ms
Reduced Features - Logistic Regression -- 	Accuracy: 0.748 / Precision: 0.381 / Recall: 0.14 / Latency: 1.0ms


#### C) Supply Vector Machine

In [90]:
# Read in models
models = {}

for mdl in ['cat', 'num', 'all', 'red']:
    models[mdl] = joblib.load('SVM_model_{}.pkl'.format(mdl))

In [91]:
# Evaluate all of our models on the validation set
evaluate_model('Categorical Features - Supply Vector Machine', models['cat'], val_features_cat, val_labels)
evaluate_model('Numerical Features - Supply Vector Machine', models['num'], val_features_num, val_labels)
evaluate_model('All Features - Supply Vector Machine', models['all'], val_features_all, val_labels)
evaluate_model('Reduced Features - Supply Vector Machine', models['red'], val_features_red, val_labels)

Categorical Features - Supply Vector Machine -- 	Accuracy: 0.768 / Precision: 0.0 / Recall: 0.0 / Latency: 32.6ms
Numerical Features - Supply Vector Machine -- 	Accuracy: 0.748 / Precision: 0.419 / Recall: 0.228 / Latency: 25.1ms
All Features - Supply Vector Machine -- 	Accuracy: 0.764 / Precision: 0.467 / Recall: 0.123 / Latency: 11.0ms
Reduced Features - Supply Vector Machine -- 	Accuracy: 0.768 / Precision: 0.0 / Recall: 0.0 / Latency: 45.2ms


C:\Users\student\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1318: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 due to no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\student\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1318: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 due to no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


#### D) Multilayer Perceptron

In [92]:
# Read in models
models = {}

for mdl in ['cat', 'num', 'all', 'red']:
    models[mdl] = joblib.load('MLP_model_{}.pkl'.format(mdl))

In [93]:
# Evaluate all of our models on the validation set
evaluate_model('Categorical Features - Multilayer Perceptron', models['cat'], val_features_cat, val_labels)
evaluate_model('Numerical Features - Multilayer Perceptron', models['num'], val_features_num, val_labels)
evaluate_model('All Features - Multilayer Perceptron', models['all'], val_features_all, val_labels)
evaluate_model('Reduced Features - Multilayer Perceptron', models['red'], val_features_red, val_labels)

Categorical Features - Multilayer Perceptron -- 	Accuracy: 0.736 / Precision: 0.214 / Recall: 0.053 / Latency: 4.0ms
Numerical Features - Multilayer Perceptron -- 	Accuracy: 0.78 / Precision: 0.714 / Recall: 0.088 / Latency: 1.1ms
All Features - Multilayer Perceptron -- 	Accuracy: 0.724 / Precision: 0.333 / Recall: 0.193 / Latency: 3.5ms
Reduced Features - Multilayer Perceptron -- 	Accuracy: 0.748 / Precision: 0.381 / Recall: 0.14 / Latency: 2.0ms


#### E) Gradient Boosting

In [94]:
# Read in models
models = {}

for mdl in ['cat', 'num', 'all', 'red']:
    models[mdl] = joblib.load('GB_model_{}.pkl'.format(mdl))

In [95]:
# Evaluate all of our models on the validation set
evaluate_model('Categorical Features - Gradient Boosting', models['cat'], val_features_cat, val_labels)
evaluate_model('Numerical Features - Gradient Boosting', models['num'], val_features_num, val_labels)
evaluate_model('All Features - Gradient Boosting', models['all'], val_features_all, val_labels)
evaluate_model('Reduced Features - Gradient Boosting', models['red'], val_features_red, val_labels)

Categorical Features - Gradient Boosting -- 	Accuracy: 0.764 / Precision: 0.0 / Recall: 0.0 / Latency: 3.0ms
Numerical Features - Gradient Boosting -- 	Accuracy: 0.756 / Precision: 0.444 / Recall: 0.211 / Latency: 1.1ms
All Features - Gradient Boosting -- 	Accuracy: 0.756 / Precision: 0.435 / Recall: 0.175 / Latency: 1.9ms
Reduced Features - Gradient Boosting -- 	Accuracy: 0.76 / Precision: 0.429 / Recall: 0.105 / Latency: 3.0ms


### Step #17: Evaluate Models on Test Set


In [106]:
# Read in our test features - categorical
test_features_cat = pd.read_csv('test_features_cat.csv')
test_labels = pd.read_csv('test_labels.csv')

In [107]:
# Read in our test features - numerical
test_features_num = pd.read_csv('test_features_num.csv')
test_labels = pd.read_csv('test_labels.csv')

In [108]:
# Read in our test features - all
test_features_all = pd.read_csv('test_features_all.csv')
test_labels = pd.read_csv('test_labels.csv')

In [109]:
# Read in our test features - reduced
test_features_red = pd.read_csv('test_features_reduced.csv')
test_labels = pd.read_csv('test_labels.csv')

#### Random Forest

In [111]:
# Read in models
models = {}

for mdl in ['cat', 'num', 'all', 'red']:
    models[mdl] = joblib.load('rf_features_{}.pkl'.format(mdl))

In [113]:
# Evaluate our final model on the test set
evaluate_model('Categorical Features', models['cat'], test_features_cat, test_labels)
evaluate_model('Numerical Features', models['num'], test_features_num, test_labels)
evaluate_model('All Features', models['all'], test_features_all, test_labels)
evaluate_model('Reduced Features', models['red'], test_features_red, test_labels)

Categorical Features -- 	Accuracy: 0.81 / Precision: 0.0 / Recall: 0.0 / Latency: 5.0ms
Numerical Features -- 	Accuracy: 0.814 / Precision: 0.5 / Recall: 0.13 / Latency: 9.4ms
All Features -- 	Accuracy: 0.822 / Precision: 0.667 / Recall: 0.087 / Latency: 6.1ms
Reduced Features -- 	Accuracy: 0.81 / Precision: 0.4 / Recall: 0.043 / Latency: 8.1ms


#### Logistic Regression

In [114]:
# Read in models
models = {}

for mdl in ['cat', 'num', 'all', 'red']:
    models[mdl] = joblib.load('LR_model_{}.pkl'.format(mdl))

In [115]:
evaluate_model('Categorical Features', models['cat'], test_features_cat, test_labels)
evaluate_model('Numerical Features', models['num'], test_features_num, test_labels)
evaluate_model('All Features', models['all'], test_features_all, test_labels)
evaluate_model('Reduced Features', models['red'], test_features_red, test_labels)

Categorical Features -- 	Accuracy: 0.818 / Precision: 0.6 / Recall: 0.065 / Latency: 2.0ms
Numerical Features -- 	Accuracy: 0.798 / Precision: 0.333 / Recall: 0.087 / Latency: 1.0ms
All Features -- 	Accuracy: 0.818 / Precision: 0.545 / Recall: 0.13 / Latency: 1.0ms
Reduced Features -- 	Accuracy: 0.818 / Precision: 0.524 / Recall: 0.239 / Latency: 1.0ms


#### Supply Vector Machine

In [116]:
# Read in models
models = {}

for mdl in ['cat', 'num', 'all', 'red']:
    models[mdl] = joblib.load('SVM_model_{}.pkl'.format(mdl))

In [117]:
evaluate_model('Categorical Features', models['cat'], test_features_cat, test_labels)
evaluate_model('Numerical Features', models['num'], test_features_num, test_labels)
evaluate_model('All Features', models['all'], test_features_all, test_labels)
evaluate_model('Reduced Features', models['red'], test_features_red, test_labels)

Categorical Features -- 	Accuracy: 0.814 / Precision: 0.0 / Recall: 0.0 / Latency: 35.0ms
Numerical Features -- 	Accuracy: 0.789 / Precision: 0.35 / Recall: 0.152 / Latency: 21.0ms
All Features -- 	Accuracy: 0.81 / Precision: 0.444 / Recall: 0.087 / Latency: 8.2ms
Reduced Features -- 	Accuracy: 0.814 / Precision: 0.0 / Recall: 0.0 / Latency: 24.0ms


C:\Users\student\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1318: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 due to no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\student\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1318: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 due to no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


#### Multilayer Perceptron

In [118]:
# Read in models
models = {}

for mdl in ['cat', 'num', 'all', 'red']:
    models[mdl] = joblib.load('MLP_model_{}.pkl'.format(mdl))

In [119]:
evaluate_model('Categorical Features', models['cat'], test_features_cat, test_labels)
evaluate_model('Numerical Features', models['num'], test_features_num, test_labels)
evaluate_model('All Features', models['all'], test_features_all, test_labels)
evaluate_model('Reduced Features', models['red'], test_features_red, test_labels)

Categorical Features -- 	Accuracy: 0.806 / Precision: 0.417 / Recall: 0.109 / Latency: 3.0ms
Numerical Features -- 	Accuracy: 0.789 / Precision: 0.25 / Recall: 0.065 / Latency: 1.0ms
All Features -- 	Accuracy: 0.798 / Precision: 0.429 / Recall: 0.261 / Latency: 1.5ms
Reduced Features -- 	Accuracy: 0.814 / Precision: 0.5 / Recall: 0.283 / Latency: 20.3ms


#### Gradient Boosting

In [120]:
# Read in models
models = {}

for mdl in ['cat', 'num', 'all', 'red']:
    models[mdl] = joblib.load('GB_model_{}.pkl'.format(mdl))

In [121]:
evaluate_model('Categorical Features', models['cat'], test_features_cat, test_labels)
evaluate_model('Numerical Features', models['num'], test_features_num, test_labels)
evaluate_model('All Features', models['all'], test_features_all, test_labels)
evaluate_model('Reduced Features', models['red'], test_features_red, test_labels)

Categorical Features -- 	Accuracy: 0.814 / Precision: 0.5 / Recall: 0.043 / Latency: 3.4ms
Numerical Features -- 	Accuracy: 0.798 / Precision: 0.423 / Recall: 0.239 / Latency: 1.0ms
All Features -- 	Accuracy: 0.806 / Precision: 0.438 / Recall: 0.152 / Latency: 4.0ms
Reduced Features -- 	Accuracy: 0.81 / Precision: 0.462 / Recall: 0.13 / Latency: 5.1ms


### Results

Model with Highest Accuracy: 0.822 - Random Forest w/ All Features

Model with Highest Precision: 0.667 - Random Forest w/ All Features

Model with Highest Recall: 0.283 - Multilayer Perceptron w/ Reduced Features

Model with Lowest Latency: Multiple w/ 1.0 ms